Lib/test/test_asyncio/test_server.py

```diff
@@ -223,28 +224,35 @@ async def serve(rd, wr):
         self.addCleanup(srv.close)
 
         addr = srv.sockets[0].getsockname()
-        (c_rd, c_wr) = await asyncio.open_connection(addr[0], addr[1])
+        (c_rd, c_wr) = await asyncio.open_connection(addr[0], addr[1], limit=4096)
         self.addCleanup(c_wr.close)
 
-        # Make sure both sides are in a paused state
-        while (s_wr.transport.get_write_buffer_size() == 0 or
-               c_wr.transport.is_reading()):
-            while s_wr.transport.get_write_buffer_size() == 0:
-                s_wr.write(b'a' * 65536)
-                await asyncio.sleep(0)
-            await asyncio.sleep(0.1) # FIXME: More socket buffer space magically appears?
+        # Limit the send buffer so we can reliably overfill it
+        s_sock = s_wr.get_extra_info('socket')
+        s_sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 65536)
+
+        # Get the reader in to a paused state by sending more than twice
+        # the configured limit
+        s_wr.write(b'a' * 4096)
+        s_wr.write(b'a' * 4096)
+        s_wr.write(b'a' * 4096)
+        while c_wr.transport.is_reading():
+            await asyncio.sleep(0)
+
+        # Get the writer in a waiting state by sending data until the
+        # kernel stops accepting more in to the send buffer
+        while s_wr.transport.get_write_buffer_size() == 0:
+            s_wr.write(b'a' * 4096)
+            await asyncio.sleep(0)
```

Comment:
Can we put an upper bound on these `while` loops and fail the test when they loop too many times? (Possibly a `for` loop with a `break` is the best idiom to do this.)

I worry that something in the network stack might cause one or the other to loop forever, and I'd rather not waste the CPU time in CI over this.

How many iterations do you expect? Is it deterministic regardless of platform?